File: sorting/quick_sort.cc
#include <iostream>

void PrintArray(const int* arr, int n) {
  std::cout << "{";
  for (int i = 0; i < n; i++)
    std::cout << arr[i] << ",";
  std::cout << "}" << std::endl;
}

int Partition(int* arr, int start, int end) {
  int pivot = arr[end];
  int p_index = start;
  for(int i = start; i < end; i++) {
    if (arr[i] <= pivot) {
      int temp = arr[i];
      arr[i] = arr[p_index];
      arr[p_index] = temp;
      p_index++;
    }
  }
  arr[end] = arr[p_index];
  arr[p_index] = pivot;
  return p_index;
}

void QuickSort(int* arr, int start, int end) {
  if (start >= end) return;
  int p_index = Partition(arr, start, end);
  QuickSort(arr, start, p_index - 1);
  QuickSort(arr, p_index + 1, end);
}

int main() {
  int n = 5;
  int* arr = new int[n];
  arr[0] = 3;
  arr[1] = 5;
  arr[2] = 1;
  arr[3] = 9;
  arr[4] = 2;
  
  PrintArray(arr, n);
  QuickSort(arr, 0, 4);
  PrintArray(arr, n);
}

File: python/quick_sort.py
# partition
def partition(arr, start, end):
  pivot = arr[end]
  p_index = start
  
  for i in range(start, end+1):
    if arr[i] < pivot:
      arr[i], arr[p_index] = arr[p_index], arr[i]
      p_index += 1
  
  arr[end], arr[p_index] = arr[p_index], arr[end]
  return p_index

# quick sort
def quick_sort(arr, start, end):
  if start >= end:
    return None
  
  p_index = partition(arr, start, end)
  quick_sort(arr, start, p_index-1)
  quick_sort(arr, p_index+1, end)

#driver program
input = [3, 5, 1, 9, 2]
print(input)

quick_sort(input, 0, len(input)-1)
print(input)